### DISCUSSION TITLE
Test method layout

### DISCUSSION THREAD

**Discussion by eNeRGy164** [fluentassertions/fluentassertions#1555] [General]
Writing 56 StreamAssertions tests and going through the 364 TypeAssertions tests, I try to get consistent code layout that fits the project.
But not very surprisingly, with 4762 tests by several authors, there is no real consistency.

I tend to go where I find a good sample to copy, and try to stick with that. But this could make things worse as I might replicate a bad example.

But for me it would be nice if there was a more _"official"_ guide how you would like to see methods to be structured.

Of course, this comes into the territory of taste, but as maintainers, you can set some guardrails so it's easier to contribute.

I can imagine the current rules might have changed over time, or that there is just a relaxed attitude.

Some things I see implemented differently:

## Variable declarations
Explicit vs. `var`

My preference is to `var` everything, I see there are sometimes SA/CA suggestions.
I think `MemberInfo memberInfo =` does not add any value over `var memberInfo =`  other then filling more space in my IDE.

using `var` also prevents the IDE from suggesting this:
```csharp
var expected = new DateTime(2016, 06, 04);
DateTime subject = new(2016, 06, 04);
```

## Inline A/A/A

I came across this example:

```csharp
// Arrange / Act / Assert
typeof(Sealed).Should().BeSealed();
```
I rather split it, put the act in a `Action` and test explicitly for not throwing an exception.

## Arrange: Explicit variables vs. inline values

I see a difference in arranging the test.

### Very explicit declaration of every parameter
```csharp
// Arrange
var type = typeof(TypeWithConversionOperators);
var sourceType = typeof(TypeWithConversionOperators);
var targetType = typeof(byte);

// Act
Action act = () =>
    type.Should().NotHaveExplicitConversionOperator(sourceType, targetType);
```

### Subject in _arrange_, values inline
```csharp
// Arrange
var type = typeof(TypeWithConversionOperators);

// Act
Action act = () =>
    type.Should().NotHaveExplicitConversionOperator(null, typeof(string));
```

I'm not really seeing a value to declare the values of the method separately. Again it takes up screen estate and they are never used for anything else. So my preference is the second one.

## Act: Action lambdas

These are written in many variations.

### single line
```csharp
// Act
Action act = () => someCollection.Should().NotBeNull();
```

### break after arrow
```csharp
// Act
Action act = () =>
    type.Should().HaveExplicitMethod(typeof(IExplicitInterface), "ExplicitMethod", null);
```

### wrap everyting
```csharp
// Act
Action act = () =>
    type.Should()
        .HaveExplicitMethod(interfaceType, "NonExistentProperty", new Type[0]);
```

### Create a straircase
```csharp
// Act
Action act = () =>
    type.Should()
        .HaveDefaultConstructor()
        .Which.Should()
            .HaveAccessModifier(CSharpAccessModifier.Public);
```

For efficient screen use, I would try to keep it on one line, or as much as to the left as possible

1. Is it a single `Should()` statement -> single line
2. Is it a single, but long line, newline after arrow
3. If it is multistatement (Which, WithXxx, And), newline after arrow, every next branch on new line

Taking the last example
```csharp
// Act
Action act = () =>
    type.Should().HaveDefaultConstructor()
        .Which.Should().HaveAccessModifier(CSharpAccessModifier.Public);
```

It's also an option to always newline after the arrow for consitency.

## Assertions

Assertions follow some unwritten rules pretty much al over the place

```csharp
// Act
// Assert
action.Should().Throw<XunitException>()
    .WithMessage("*at least*4*because we want to test the failure message*3*");
```

1. Single line
2. Or after `[Not]Throw` indented lines

**Comment by dennisdoomen** [comment]
In general we follow the https://csharpcodingguidelines.com. We also planned to capture certain design decisions and conventions in https://github.com/fluentassertions/fluentassertions/issues/1340 

> Explicit vs. `var`

There's no simple rule here. But every relationship has two aspects: the role and the type. If the type is clear from the statement or really not important (which is subjective), use `var`. If it not, be explicit. 



> I rather split it, put the act in a `Action` and test explicitly for not throwing an exception.

We've always done that. But recently, I started to think that this may be overkill. If the test throws unexpectedly, the test would fail anyway. 

> 


> ### Subject in _arrange_, values inline
> ```cs
> // Arrange
> var type = typeof(TypeWithConversionOperators);
> 
> // Act
> Action act = () =>
>     type.Should().NotHaveExplicitConversionOperator(null, typeof(string));
> ```
> 
> I'm not really seeing a value to declare the values of the method separately. Again it takes up screen estate and they are never used for anything else. So my preference is the second one.

In general I prefer to declare the subject explicitly, but maybe we're being overzealous sometimes. 

> ## Act: Action lambdas
> These are written in many variations.

I have no specific preference here. I just rely on the `.editorconfig` to reformat based on the max-130-character-rule.

> Assertions follow some unwritten rules pretty much al over the place
> 
> ```cs
> // Act
> // Assert
> action.Should().Throw<XunitException>()
>     .WithMessage("*at least*4*because we want to test the failure message*3*");
> ```
> 
> 1. Single line
> 2. Or after `[Not]Throw` indented lines

Same here. Just stay under 130 characters.

**Reply by eNeRGy164** [reply]
I should test the 130 char rule then, didn't see it showing up yet

**Reply by eNeRGy164** [reply]
It's R# rule, it is in the .DotSettings file

So it won't be applied in my IDE.

-- Edit --

There is a PR for now 🤡  #1556 

-- Edit 2 --

Apperantly both VS & VS Code do not support it, will ook how to implement the wrapping,

**Reply by dennisdoomen** [reply]
I was about to say that we never added `max_line_length` to the `.editorconfig` ;-)
